### ISSUE TITLE
Fix Database Schema

### ISSUE THREAD

**Issue by 0x4007** [ubiquibot/generate-vector-embeddings#8]
This requires changes to the plugin that captures the data, as well as the database itself. 

![2024-09-08 21 54 35](https://github.com/user-attachments/assets/b6885a88-accb-4349-bf39-bb96b890488c)

This must capture issue bodies. 

---

Perhaps I wasn't explicit with my vision for the database schema, but this needs adjustments.

We only need: 

```
id, created_at, modified_at, author_id, plaintext, embedding
```

- The `plaintext` should only save if its a public repo, otherwise we have liabilities on our hands storing company private data. 
- We can make a separate table for when we start capturing messages from Telegram and other platforms. No concerns for now. 
- ID might need to be the GraphQL node id, because, as I understand, the GitHub REST API will not allow us to query the comment with only the numerical ID. We would need to store the organization name, and the repository name, before we can query the comment based on its ID. 
- Dates: the reason for capturing the dates is because I envision that we can make a Telegram chatbot that will allow anybody to query anything about the organization's latest projects in plain English. This is useful for onboarding (new team member has some questions about our infrastructure/products) and investors wanting an update. We need to give less weight to older comments. 
- author id: for onboarding developers especially, it would be useful to target specific contributors' inputs. Mine for final decisions, and others who may be the primary contributors to specific repos, compared to other contributors etc.

**Comment by sshivaditya2019** [comment]
@0x4007 would plaintext include issue body as well or would it just be comment body?

So, right now if there's a private repository then the plaintext would become 'CENSORED'. Is that fine?

**Comment by 0x4007** [comment]
Issue body as well and no leave null

**Comment by 0x4007** [comment]
Why do you want to associate and/or combine the embedding of the issue body and the comment body?

**Comment by sshivaditya2019** [comment]
> Why do you want to associate and/or combine the embedding of the issue body and the comment body?
> 
> Can you provide some examples of applications that would require this?

Case 1: Issue Body + Comment Body:
- Only one vector index for `Issue Body + Comment Body`
- Might support vector search if we want to search for comments or issues alone.
- Uses: Relevance of the comment to the issue body, Text Based Searching using LIKE

Case 2: Issue Body and Comment Body Separate columns:
- Multiple Vector index required for vector search, as we want to support vector search by issue and comment text
- Multiple Columns (Might not be bad?)
- Uses: TLDR comments, comments/issue based vector search

If having two separate embeddings is acceptable, then that's fine. According to the issue's specification, the desired schema was `id, created_at, modified_at, author_id, plaintext, embedding`.

**Comment by 0x4007** [comment]
I don't understand your explanation. 

Anyways would be great if we can make the following applications:

1. Issue deduplication.
2. Assistant to answer questions specifically related to our DAO and products here on github and telegram in plain English

**Comment by sshivaditya2019** [comment]
> I don't understand your explanation. 
> 
> Anyways would be great if we can make the following applications:
> 
> 1. Issue deduplication.
> 2. Assistant to answer questions specifically related to our DAO and products here on github and telegram in plain English 
> 
> As I understand with my proposed schema, these applications are possible. 
> 
> I wonder if it makes sense simply to capture every property, excluding the URL properties from the issue and comment objects on GitHub 

Just to be clear, what is plaintext in your schema ?

**Comment by 0x4007** [comment]
The source code of the markdown of the comments. 

I hope that the mark down syntax doesn't negatively affect the LLM's ability to understand. 

I know that I always must clarify with ChatGPT that the right arrow bracket means it's a block quote from another comment.

**Comment by sshivaditya2019** [comment]
> The source code of the markdown of the comments.
> 
> I hope that the mark down syntax doesn't negatively affect the LLM's ability to understand.
> 
> I know that I always must clarify with ChatGPT that the right arrow bracket means it's a block quote from another comment.
> 
> I suppose we'll need to compare performance of raw marked down source code and something preprocessed/more similar to how we perceive it

I'll give an example of plaintext let me know if this right ?

```
## Issue #1: Sample Issue with Replies

This is the body of the first issue.

### Comment by user1:

This is the first comment on issue #1.

#### Reply by user2:

This is a reply to the first comment by user1.

#### Reply by user3:

This is another reply to the first comment by user1.

### Comment by user4:

This is a second comment on issue #1.
```

Is this right ? If not could you please give an example. I am not sure what you mean by ` source code of the markdown of the comments`.

**Comment by sshivaditya2019** [comment]
> I am not an expert working with embeddings but as I understand we need to generate a single embedding per corpus of text (comment)
> 
> Otherwise yes you included the markdown syntax which is what I said.

To clarify, the example I shared involves a single embedding/vector. Currently, this means that a new text corpus(markdown I shared before) will be created for each new issue, and additional comments will be added to that corpus.

Currently, we create a new record for each comment and generate an embedding for each one. This approach won't work with the issuededup or Assistant plan you have.

**Comment by 0x4007** [comment]
So then if we are clobbering a single embedding to represent an entire issue (and pull) won't the costs increase exponentially? Can't we feed multiple embeddings to an LLM to work with? 

It doesn't seem like a good approach to clobber

**Comment by sshivaditya2019** [comment]
> So then if we are clobbering a single embedding to represent an entire issue (and pull) won't the costs increase exponentially? Can't we feed multiple embeddings to an LLM to work with?
> 
> It doesn't seem like a good approach to clobber

An alternative approach would be to store each piece of text from issues/comments as separate entities and associate them with their respective node IDs. This would create a large text corpus containing all the content from the organization, which could be identified by a global node ID or by using a type column in the schema.

**Comment by sshivaditya2019** [comment]
> So then if we are clobbering a single embedding to represent an entire issue (and pull) won't the costs increase exponentially? Can't we feed multiple embeddings to an LLM to work with?
> 
> It doesn't seem like a good approach to clobber

To clarify, we don’t provide embeddings directly to the LLM. Instead, we perform a vector or similarity search, apply a ranking technique, and then extract text values that are most relevant to the search subject. This extracted text is then used as context for the LLM.

**Comment by 0x4007** [comment]
@gentlementlegen im assuming storing node id won't be sufficient for us to determine which embeddings are associated with eachother. I'm assuming we need to also store the issue id

**Comment by 0x4007** [comment]
> [What are embeddings?](https://platform.openai.com/docs/guides/embeddings/what-are-embeddings)
OpenAI’s text embeddings measure the relatedness of text strings. Embeddings are commonly used for:
> 
> Search (where results are ranked by relevance to a query string)
Clustering (where text strings are grouped by similarity)
Recommendations (where items with related text strings are recommended)
Anomaly detection (where outliers with little relatedness are identified)
Diversity measurement (where similarity distributions are analyzed)
Classification (where text strings are classified by their most similar label)

**Comment by sshivaditya2019** [comment]
> @gentlementlegen im assuming storing node id won't be sufficient for us to determine which embeddings are associated with eachother. I'm assuming we need to also store the issue id
> 
> Anyways at this point im starting to think we should just store everything besides the _url properties and then we can optimize later since we seem to not know for sure what to do now.

@0x4007 

Hi, could you update me on the final database schema? I believe that the global node ID and type should be sufficient. Other parameters might not be relevant to the embedding text corpus and could end up being more of a log dump.

**Comment by sshivaditya2019** [comment]
> > [What are embeddings?](https://platform.openai.com/docs/guides/embeddings/what-are-embeddings)
> > OpenAI’s text embeddings measure the relatedness of text strings. Embeddings are commonly used for:
> > Search (where results are ranked by relevance to a query string)
> > Clustering (where text strings are grouped by similarity)
> > Recommendations (where items with related text strings are recommended)
> > Anomaly detection (where outliers with little relatedness are identified)
> > Diversity measurement (where similarity distributions are analyzed)
> > Classification (where text strings are classified by their most similar label)
> 
> ChatGPT: Here are additional use cases for embeddings:
> 
> > 1. **Sentiment Analysis**: Embeddings can help map text to a sentiment space, allowing you to analyze the emotional tone of a text (positive, negative, neutral).
> > 2. **Language Translation**: Embeddings capture semantic meaning, making them useful for machine translation, where similar sentences in different languages are close in embedding space.
> > 3. **Information Extraction**: Identify and extract structured information from unstructured text by mapping words or phrases to relevant concepts.
> > 4. **Text Summarization**: Summarize large documents by comparing sentence embeddings and extracting the most important, representative sentences.
> > 5. **Question Answering**: Embeddings can improve the ability to match questions with the most relevant answers based on semantic similarity.
> > 6. **Text Generation**: Used in generative models like GPT, embeddings help maintain coherence in text generation by representing context.
> > 7. **Content Moderation**: Classify or filter harmful or inappropriate content by mapping texts to predefined categories of acceptable and non-acceptable behavior.
> > 8. **Topic Modeling**: Discover underlying topics within a large set of documents by clustering similar text embeddings.
> 
> Reviewing the [leaderboard](https://huggingface.co/spaces/mteb/leaderboard) and I realize that STS (Semantic Textual Similarity) is the most important for depuplication and task matchmaking. Check the STS tab. `voyage-lite-02-instruct` looks promising (the first ranked seems specifically designed for English-French translations?)
> 
> Seems significantly better than OpenAI's offering.
> 
> We may also be able to make a feature that allows contributors to ask product/strategic direction questions to the bot. It can link back to comments that answer their question.
> 
> Text summarization might be able to improve our relevance scoring of comments (verbose and concise comments may be able to receive similar "quantitative scoring" credit)

I can probably write a new adapter for them, but I think their free tier has only 3 RPM (Request per minute), but offer 50M free tokens for new accounts to use ?

**Comment by 0x4007** [comment]
Three requests per minute seems acceptable as long as we can have some type of buffer or queue. I believe we had plans to do this anyways with the OpenAI embeddings because it's significantly cheaper. 

However, can we generate multiple embeddings per request? If not then I imagine this can get more complicated than a cron job. 

As for schema, I think it's simpler to dump everything and optimize later. This is because we don't know what the next plugins will need and this gives us more flexibility for research. To me it's only a priority to optimize when we start hitting free tier limits.

**Comment by sshivaditya2019** [comment]
> Three requests per minute seems acceptable as long as we can have some type of buffer or queue. I believe we had plans to do this anyways with the OpenAI embeddings because it's significantly cheaper.
> 
> However, can we generate multiple embeddings per request? If not then I imagine this can get more complicated than a cron job.
> 
> As for schema, I think it's simpler to dump everything and optimize later. This is because we don't know what the next plugins will need and this gives us more flexibility for research. To me it's only a priority to optimize when we start hitting free tier limits.

So, if we add a payment method, the requests per minute can be increased. I have updated the plugin for support with voyageai

**Comment by 0x4007** [comment]
![image](https://github.com/user-attachments/assets/eaf5c732-5c86-415b-9a70-b5b46069cc72)

Any name and zip code should work

**Comment by sshivaditya2019** [comment]
> ![image](https://private-user-images.githubusercontent.com/4975670/366257914-eaf5c732-5c86-415b-9a70-b5b46069cc72.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwMTU1NDAsIm5iZiI6MTcyNjAxNTI0MCwicGF0aCI6Ii80OTc1NjcwLzM2NjI1NzkxNC1lYWY1YzczMi01Yzg2LTQxNWItOWE3MC1iNWI0NjA2OWNjNzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTFUMDA0MDQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODc3MmNlZDNjNmZiNjE5MTBiZmUwZTA5YmMyMmI1Y2JkNTcwZjZkZTJhZDIxOTBmYzQ0NDVjYjgxMTcwNzk2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.TkZAMXEaSwT3iBvdhv1xcrZxO8ZETCqnrGAoXo16578)
> 
> Any name and zip code should work

I've added my payment method. Please delete this information immediately or share it through another platform, like Telegram.

**Comment by 0x4007** [comment]
It's merchant locked for up to $1

**Comment by sshivaditya2019** [comment]
So, the final schema would be `Id (GlobalNodeId), plaintext(text from the comment),createdAt, ModifiedAt, embedding`
and this entire thing as a serialized object
```
comment: {
        /**
         * AuthorAssociation
         * @description How the author is associated with the repository.
         * @enum {string}
         */
        author_association:
          | "COLLABORATOR"
          | "CONTRIBUTOR"
          | "FIRST_TIMER"
          | "FIRST_TIME_CONTRIBUTOR"
          | "MANNEQUIN"
          | "MEMBER"
          | "NONE"
          | "OWNER";
        /** @description Contents of the issue comment */
        body: string;
        /** Format: date-time */
        created_at: string;
        /** Format: uri */
        html_url: string;
        /**
         * Format: int64
         * @description Unique identifier of the issue comment
         */
        id: number;
        /** Format: uri */
        issue_url: string;
        node_id: string;
        performed_via_github_app: null | components["schemas"]["integration"];
        /** Reactions */
        reactions: {
          "+1": number;
          "-1": number;
          confused: number;
          eyes: number;
          heart: number;
          hooray: number;
          laugh: number;
          rocket: number;
          total_count: number;
          /** Format: uri */
          url: string;
        };
        /** Format: date-time */
        updated_at: string;
        /**
         * Format: uri
         * @description URL for the issue comment
         */
        url: string;
        /** User */
        user: {
          /** Format: uri */
          avatar_url?: string;
          deleted?: boolean;
          email?: string | null;
          /** Format: uri-template */
          events_url?: string;
          /** Format: uri */
          followers_url?: string;
          /** Format: uri-template */
          following_url?: string;
          /** Format: uri-template */
          gists_url?: string;
          gravatar_id?: string;
          /** Format: uri */
          html_url?: string;
          id: number;
          login: string;
          name?: string;
          node_id?: string;
          /** Format: uri */
          organizations_url?: string;
          /** Format: uri */
          received_events_url?: string;
          /** Format: uri */
          repos_url?: string;
          site_admin?: boolean;
          /** Format: uri-template */
          starred_url?: string;
          /** Format: uri */
          subscriptions_url?: string;
          /** @enum {string} */
          type?: "Bot" | "User" | "Organization";
          /** Format: uri */
          url?: string;
        } | null;
      };
```

**Comment by 0x4007** [comment]
Looks like there isn't a lot to save from there if we remove the URLs, nested objects like reactions and user details.  

I'm surprised that's all that's there.

**Comment by sshivaditya2019** [comment]
> Looks like there isn't a lot to save from there if we remove the URLs, nested objects like reactions and user details.
> 
> I'm surprised that's all that's there.
> 
> Performed via GitHub app should also be removed because we shouldn't be generating embeddings for bot comments.

I took this from octokit's type definition. I think reactions should be kept, as it could be a metric of engagement for comments with low relevance.

**Comment by 0x4007** [comment]
My concern is that I feel we should make everything into its own column. However, for simplicity, we could just save the full reaction object in a single column.

**Comment by 0x4007** [comment]
I think Supabase is probably fine because there are no planned features for reactions. We unlikely need the performance for querying for a long time. Even if or when we do, we can consider adding those columns later.

**Comment by sshivaditya2019** [comment]
> I think Supabase is probably fine because there are no planned features for reactions. We unlikely need the performance for querying for a long time. Even if or when we do, we can consider adding those columns later. 
> 
> To be honest I'm not a database expert but seems like Supabase should handle our needs. It has a nice UI which is convenient for debugging

It should be better. I am skeptical about storing serialized comment objects in Postgres database. It should not affect the performance though.

**Comment by 0x4007** [comment]
I asked ChatGPT for a new feature: onboarding new developers, via plaintext q&a. It said we should generate embeddings of an entire repository. How does this affect our schema? Perhaps we should generalize further? Or we can make a new table?

**Comment by sshivaditya2019** [comment]
> I asked ChatGPT for a new feature: onboarding new developers, via plaintext q&a. It said we should generate embeddings of an entire repository. How does this affect our schema? Perhaps we should generalize further? Or we can make a new table?
> 
> Perhaps a new table per GitHub object type is the most manageable.

that would depend on the onboarding, if onboarding involves explaining the current tickets and comments and the work that's going on, this schema would be enough.

If we want to explain the code and other things, for them we do not need embeddings. For General Q&A this should be enough. We could further expand this by adding OpenAI functions, to query the repo based on the serialized object retrieved using vector search.

**Comment by 0x4007** [comment]
I'll need to top up the wallet soon
